packages/vite/src/node/utils.ts

```diff
@@ -534,15 +534,35 @@ export function isFileReadable(filename: string): boolean {
 
 /**
  * Delete every file and subdirectory. **The given directory must exist.**
- * Pass an optional `skip` array to preserve files in the root directory.
+ * Pass an optional `skip` array to preserve files under the root directory.
  */
-export function emptyDir(dir: string, skip?: string[]): void {
+export function emptyDir(dir: string, skip: string[] = []): void {
+  let nested: Map<string, string[]> | null = null
+  skip = skip.map(slash)
   for (const file of fs.readdirSync(dir)) {
-    if (skip?.includes(file)) {
+    const matched = skip.find((f) => f === file || f.startsWith(`${file}/`))
+    if (matched) {
+      if (matched !== file) {
+        nested ??= new Map()
+        let nestedSkip = nested.get(`${dir}/${file}`)
+        if (!nestedSkip) {
+          nestedSkip = []
+          nested.set(`${dir}/${file}`, nestedSkip)
+        }
+        const skipPath = matched.replace(`${file}/`, '')
+        if (!nestedSkip.includes(skipPath)) {
+          nestedSkip.push(skipPath)
+        }
```

Comment:
I think this still won't give you the correct `nestedSkip` list for each nested dir. If you have in skip: `[ 'foo/a', 'foo/b' ]`, you also need to have `b` in `nested.get('foo')`. It would be also good to avoid creating the extra `skip` array when it isn't needed. I think pre-processing the skip list as I suggested before may end up making the code easier to follow.

Reply:
Sorry, I revise it.